File: csharp/Platform.Protocols.Lino.Tests/EdgeCaseParserTests.cs
using System;
using Xunit;

namespace Platform.Protocols.Lino.Tests
{
    public static class EdgeCaseParserTests
    {
        [Fact]
        public static void EmptyLinkTest()
        {
            var source = @":";
            var parser = new Parser();
            // Standalone ':' is now forbidden and should throw an exception
            Assert.Throws<FormatException>(() => parser.Parse(source));
        }

        [Fact]
        public static void EmptyLinkWithParenthesesTest()
        {
            var source = @"()";
            var target = @"()";
            var parser = new Parser();
            var links = parser.Parse(source);
            var formattedLinks = links.Format();
            Assert.Equal(target, formattedLinks);
        }

        [Fact]
        public static void EmptyLinkWithEmptySelfReferenceTest()
        {
            var source = @"(:)";
            var parser = new Parser();
            // '(:)' is now forbidden and should throw an exception
            Assert.Throws<FormatException>(() => parser.Parse(source));
        }

        [Fact]
        public static void TestAllFeaturesTest()
        {
            // Test single-line link with id
            var input = "id: value1 value2";
            var result = new Parser().Parse(input);
            Assert.NotEmpty(result);

            // Test multi-line link with id
            input = "(id: value1 value2)";
            result = new Parser().Parse(input);
            Assert.NotEmpty(result);

            // Test link without id (single-line) - now forbidden
            input = ": value1 value2";
            Assert.Throws<FormatException>(() => new Parser().Parse(input));

            // Test link without id (multi-line) - now forbidden
            input = "(: value1 value2)";
            Assert.Throws<FormatException>(() => new Parser().Parse(input));

            // Test singlet link
            input = "(singlet)";
            result = new Parser().Parse(input);
            Assert.NotEmpty(result);

            // Test value link
            input = "(value1 value2 value3)";
            result = new Parser().Parse(input);
            Assert.NotEmpty(result);

            // Test quoted references
            input = @"(""id with spaces"": ""value with spaces"")";
            result = new Parser().Parse(input);
            Assert.NotEmpty(result);

            // Test single-quoted references
            input = "('id': 'value')";
            result = new Parser().Parse(input);
            Assert.NotEmpty(result);

            // Test nested links
            input = "(outer: (inner: value))";
            result = new Parser().Parse(input);
            Assert.NotEmpty(result);
        }

        [Fact]
        public static void TestSingletLinks()
        {
            // Test singlet (1)
            var input = "(1)";
            var result = new Parser().Parse(input);
            Assert.Single(result);
            Assert.Equal("1", result[0].Id);
            Assert.Null(result[0].Values);

            // Test (1 2)
            input = "(1 2)";
            result = new Parser().Parse(input);
            Assert.Single(result);
            Assert.Null(result[0].Id);
            Assert.NotNull(result[0].Values);
            Assert.Equal(2, result[0].Values.Count);
            Assert.Equal("1", result[0].Values[0].Id);
            Assert.Equal("2", result[0].Values[1].Id);

            // Test (1 2 3)
            input = "(1 2 3)";
            result = new Parser().Parse(input);
            Assert.Single(result);
            Assert.Null(result[0].Id);
            Assert.NotNull(result[0].Values);
            Assert.Equal(3, result[0].Values.Count);
            Assert.Equal("1", result[0].Values[0].Id);
            Assert.Equal("2", result[0].Values[1].Id);
            Assert.Equal("3", result[0].Values[2].Id);

            // Test (1 2 3 4)
            input = "(1 2 3 4)";
            result = new Parser().Parse(input);
            Assert.Single(result);
            Assert.Null(result[0].Id);
            Assert.NotNull(result[0].Values);
            Assert.Equal(4, result[0].Values.Count);
            Assert.Equal("1", result[0].Values[0].Id);
            Assert.Equal("2", result[0].Values[1].Id);
            Assert.Equal("3", result[0].Values[2].Id);
            Assert.Equal("4", result[0].Values[3].Id);
        }

        [Fact]
        public static void TestEmptyDocumentTest()
        {
            var input = "";
            // Empty document should return empty list
            var result = new Parser().Parse(input);
            Assert.Empty(result);
        }

        [Fact]
        public static void TestWhitespaceOnlyTest()
        {
            var input = "   \n   \n   ";
            // Whitespace-only document should return empty list (similar to empty document)
            var result = new Parser().Parse(input);
            Assert.Empty(result);
        }

        [Fact]
        public static void TestEmptyLinksTest()
        {
            var input = "()";
            var result = new Parser().Parse(input);
            Assert.NotEmpty(result);
            
            // '(:)' is now forbidden
            input = "(:)";
            Assert.Throws<FormatException>(() => new Parser().Parse(input));
            
            input = "(id:)";
            result = new Parser().Parse(input);
            Assert.NotEmpty(result);
        }
    }
}

File: rust/tests/edge_case_parser_tests.rs
use ln::{parse_ln, parse_ln_to_links, Ln};

#[test]
fn empty_link_test() {
    let source = ":";
    // Standalone ':' is now forbidden and should return an error
    let result = parse_ln(source);
    assert!(result.is_err());
}

#[test]
fn empty_link_with_parentheses_test() {
    let source = "()";
    let parsed = parse_ln(source).unwrap();
    // Should parse as an empty link
    assert!(parsed.is_link());
    // Verify formatted output matches C#/JS
    let output = format!("{:#}", parsed);
    assert_eq!("()", output);
}

#[test]
fn empty_link_with_empty_self_reference_test() {
    let source = "(:)";
    // '(:)' is now forbidden and should return an error
    let result = parse_ln(source);
    assert!(result.is_err());
}

#[test]
fn test_all_features_test() {
    // Test single-line link with id
    let input = "id: value1 value2";
    let result = parse_ln(input);
    assert!(result.is_ok());

    // Test multi-line link with id
    let input = "(id: value1 value2)";
    let result = parse_ln(input);
    assert!(result.is_ok());

    // Test link without id (single-line) - now forbidden
    let input = ": value1 value2";
    let result = parse_ln(input);
    assert!(result.is_err());

    // Test link without id (multi-line) - now forbidden
    let input = "(: value1 value2)";
    let result = parse_ln(input);
    assert!(result.is_err());

    // Test singlet link
    let input = "(singlet)";
    let result = parse_ln(input);
    assert!(result.is_ok());
    let parsed = result.unwrap();
    if let Ln::Link { id, values } = parsed {
        assert!(id.is_none());
        assert_eq!(values.len(), 1);
        if let Ln::Ref(ref_id) = &values[0] {
            assert_eq!(ref_id, "singlet");
        }
    }

    // Test value link
    let input = "(value1 value2 value3)";
    let result = parse_ln(input);
    assert!(result.is_ok());

    // Test quoted references
    let input = r#"("id with spaces": "value with spaces")"#;
    let result = parse_ln(input);
    assert!(result.is_ok());

    // Test single-quoted references
    let input = "('id': 'value')";
    let result = parse_ln(input);
    assert!(result.is_ok());

    // Test nested links
    let input = "(outer: (inner: value))";
    let result = parse_ln(input);
    assert!(result.is_ok());
}

#[test]
fn test_empty_document_test() {
    let input = "";
    // Empty document should return empty list (matching C#/JS behavior)
    let result = parse_ln_to_links(input);
    assert!(result.is_ok());
    let parsed = result.unwrap();
    assert!(parsed.is_empty());
}

#[test]
fn test_whitespace_only_test() {
    let input = "   \n   \n   ";
    // Whitespace-only document should return empty list (matching C#/JS behavior)
    let result = parse_ln_to_links(input);
    assert!(result.is_ok());
    let parsed = result.unwrap();
    assert!(parsed.is_empty());
}

#[test]
fn test_empty_links_test() {
    let input = "()";
    let result = parse_ln(input);
    assert!(result.is_ok());
    
    // '(:)' is now forbidden
    let input = "(:)";
    let result = parse_ln(input);
    assert!(result.is_err());
    
    let input = "(id:)";
    let result = parse_ln(input);
    assert!(result.is_ok());
}

#[test]
fn test_singlet_links() {
    // Test singlet (1)
    let input = "(1)";
    let result = parse_ln(input);
    assert!(result.is_ok());
    let parsed = result.unwrap();
    if let Ln::Link { id, values } = parsed {
        assert!(id.is_none());
        assert_eq!(values.len(), 1);
        if let Ln::Ref(ref_id) = &values[0] {
            assert_eq!(ref_id, "1");
        }
    }

    // Test (1 2)
    let input = "(1 2)";
    let result = parse_ln(input);
    assert!(result.is_ok());
    let parsed = result.unwrap();
    if let Ln::Link { id, values } = parsed {
        assert!(id.is_none());
        assert_eq!(values.len(), 1);
        if let Ln::Link { id, values } = &values[0] {
            assert!(id.is_none());
            assert_eq!(values.len(), 2);
            assert_eq!(values[0], Ln::Ref("1".to_string()));
            assert_eq!(values[1], Ln::Ref("2".to_string()));
        }
    }

    // Test (1 2 3)
    let input = "(1 2 3)";
    let result = parse_ln(input);
    assert!(result.is_ok());
    let parsed = result.unwrap();
    if let Ln::Link { id, values } = parsed {
        assert!(id.is_none());
        assert_eq!(values.len(), 1);
        if let Ln::Link { id, values } = &values[0] {
            assert!(id.is_none());
            assert_eq!(values.len(), 3);
            assert_eq!(values[0], Ln::Ref("1".to_string()));
            assert_eq!(values[1], Ln::Ref("2".to_string()));
            assert_eq!(values[2], Ln::Ref("3".to_string()));
        }
    }

    // Test (1 2 3 4)
    let input = "(1 2 3 4)";
    let result = parse_ln(input);
    assert!(result.is_ok());
    let parsed = result.unwrap();
    if let Ln::Link { id, values } = parsed {
        assert!(id.is_none());
        assert_eq!(values.len(), 1);
        if let Ln::Link { id, values } = &values[0] {
            assert!(id.is_none());
            assert_eq!(values.len(), 4);
            assert_eq!(values[0], Ln::Ref("1".to_string()));
            assert_eq!(values[1], Ln::Ref("2".to_string()));
            assert_eq!(values[2], Ln::Ref("3".to_string()));
            assert_eq!(values[3], Ln::Ref("4".to_string()));
        }
    }
}

#[test]
fn test_invalid_input() {
    let input = "(invalid";
    let result = parse_ln(input);
    assert!(result.is_err());
}